File: src/08_Amortized_Analysis/nearest_smaller_element.cpp
#include <bits/stdc++.h>
#include "../compete.hpp"
using namespace std;


vector<int> v{1, 3, 4, 2, 5, 3, 4, 2, 0, 1};


// amortized O(n)
void solve()
{
  cout << v << '\n';
  stack<int> s;

  for (auto c : v) {
    while ((!s.empty()) && (s.top() >= c)) s.pop();
    if (!s.empty()) cout << s.top() << ' ';
    else cout << "- ";
    s.push(c);
  }
  cout << '\n';
}


int main() {solve();}

File: src/08_Amortized_Analysis/nearest_smaller_element.py
v = [1, 3, 4, 2, 5, 3, 4, 2, 0, 1]


# amortized O(n)
def solve():
  stack = []
  ans = []

  for x in v:
    while (stack) and (stack[-1] >= x):
      stack.pop()
    if stack:
      ans.append(stack[-1])
    else:
      ans.append('-')
    stack.append(x)

  print(' '.join(map(str, v)))
  print(' '.join(map(str, ans)))


if __name__ == '__main__':
  solve()
